### PR TITLE
feat[api]: new get user profile end point

### PR DESCRIPTION
**Pull request by theus77**
|Q              |A  |
|---------------|---|
|Bug fix?       |N|
|New feature?   |Y|	
|BC breaks?     |N|	
|Deprecations?  |N|	
|Fixed tickets  |N|	

Get, via the rest api, the username, the display name, the email and the roles of the user

### CODE REVIEW

File: Controller/ContentManagement/CrudController.php [hunk 1]
```diff
@@ -358,4 +359,25 @@ public function getContentTypeInfo(ContentType $contentType)
                 'contentType' => $contentType,
         ]);
     }
+
+    /**
+     * @Route("/api/user-profile", defaults={"_format": "json"}, methods={"GET"})
+     */
+    public function getUserProfile()
```
Comment (coppee) [inline Controller/ContentManagement/CrudController.php:366]: Adding the return type ?

File: Controller/ContentManagement/CrudController.php [hunk 2]
```diff
@@ -358,4 +361,46 @@ public function getContentTypeInfo(ContentType $contentType)
...
+            'roles' => $user->getRoles(),
+            'email' => $user->getEmail(),
+            'circles' => $user->getCircles(),
+        ]);
+    }
+
+    /**
+     * @Route("/api/user-profiles", defaults={"_format": "json"}, methods={"GET"})
+     */
+    public function getUserProfiles(UserService $userService) : JsonResponse
```
Comment (coppee) [inline Controller/ContentManagement/CrudController.php:390]: Question : peut-on "autowirer" un service dans un bundle ?

Comment (theus77) [reply Controller/ContentManagement/CrudController.php:390]: Non c'est vrai. J'ai été trop vite

Comment (theus77) [reply Controller/ContentManagement/CrudController.php:390]: @Davidmattei what do you think about Jérome's question? How should we attach services to controller in the core bundle

Comment (Davidmattei) [reply Controller/ContentManagement/CrudController.php:390]: Inject them in the constructor, autowiring in third party bundles should not be done. It makes overwriting them harder.

Services should not use autowiring or autoconfiguration. Instead, all services should be defined explicitly._

https://symfony.com/doc/current/bundles/best_practices.html#services

Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: Fixed cfe5cdb5f353154596abff458eebe79179ee0cec

File: Controller/ContentManagement/CrudController.php [hunk 3]
```diff
@@ -358,4 +374,47 @@ public function getContentTypeInfo(ContentType $contentType)
                 'contentType' => $contentType,
         ]);
     }
+
+    /**
+     * @Route("/api/user-profile", defaults={"_format": "json"}, methods={"GET"})
+     */
+    public function getUserProfile() : JsonResponse
```
Comment (dameert) [inline Controller/ContentManagement/CrudController.php:381]:
```suggestion
    public function getUserProfile(): JsonResponse
```

We should get this reported by phpCS!

File: Controller/ContentManagement/CrudController.php [hunk 4]
```diff
@@ -358,4 +374,47 @@ public function getContentTypeInfo(ContentType $contentType)
...
+            throw new \RuntimeException('User disabled');
+        }
+
+        return $this->json([
+            'username' => $user->getUsername(),
+            'displayName' => $user->getDisplayName(),
+            'roles' => $user->getRoles(),
+            'email' => $user->getEmail(),
+            'circles' => $user->getCircles(),
+        ]);
```
Comment (dameert) [inline Controller/ContentManagement/CrudController.php:397]: This conversion to json is calling many functions on $user, a flat controller should only be the glue. Can the array of information for the ->json call be generated in the $user object?

```suggestion
        return $this->json($user->toArray());
```

File: Controller/ContentManagement/CrudController.php [hunk 5]
```diff
@@ -358,4 +374,47 @@ public function getContentTypeInfo(ContentType $contentType)
...
+    /**
+     * @Route("/api/user-profiles", defaults={"_format": "json"}, methods={"GET"})
+     * @IsGranted("ROLE_ADMIN")
+     */
+    public function getUserProfiles() : JsonResponse
+    {
+        $users = [];
+        foreach ($this->userService->getAllUsers() as $user) {
+            if ($user->isEnabled()) {
+                $users[] = [
```
Comment (dameert) [inline Controller/ContentManagement/CrudController.php:409]: aha the ->toArray() function can allready be re-used here!

File: Controller/ContentManagement/CrudController.php [hunk 6]
```diff
@@ -358,4 +374,47 @@ public function getContentTypeInfo(ContentType $contentType)
...
+                    'username' => $user->getUsername(),
+                    'displayName' => $user->getDisplayName(),
+                    'roles' => $user->getRoles(),
+                    'email' => $user->getEmail(),
+                    'circles' => $user->getCircles(),
+                ];
+            }
+        }
+        return $this->json($users);
```
Comment (dameert) [inline Controller/ContentManagement/CrudController.php:418]: same comment, the controller should be the glue.

```suggestion
        return $this->json($this->userService->getEnabledUsers());
```

File: Entity/User.php [hunk 7]
```diff
@@ -199,4 +199,16 @@ public function isEnabled();
      * @return string
      */
     public function getUsername();
+
+
+    /**
+     * @return array
+     */
+    public function getRoles();
```
Comment (dameert) [inline Entity/User.php:207]:
```suggestion
    public function getRoles(): array;
```
Is this notation not supported for entities?

File: Controller/ContentManagement/CrudController.php (hunk 4 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php:397]: If $user was refering to a class I would agree with you, but here it's an interface. So a method toArray might refers to a serialization method and so expose sensitive information in the api

Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: I centralized the User to array conversion : 3d0a7d37515d4683568aae7b957bc3477a93cda7 A full refactoring of controller should be done and mainly refers to an ApiService

File: Controller/ContentManagement/CrudController.php (hunk 5 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: 3d0a7d37515d4683568aae7b957bc3477a93cda7

File: Controller/ContentManagement/CrudController.php (hunk 6 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: To me the $this->userService->getEnabledUsers() could expose too much info, we should have something like $this->apiService->getUsers()? What do you think?

File: Entity/User.php [hunk 8]
```diff
@@ -199,4 +199,16 @@ public function isEnabled();
      * @return string
      */
     public function getUsername();
+
+
+    /**
+     * @return array<string>
```
Comment (Davidmattei) [inline Entity/User.php:205]: 👍

File: Controller/ContentManagement/README.md [hunk 9]
```diff
@@ -0,0 +1,14 @@
+Using the rest API
```
Comment (Davidmattei) [inline Controller/ContentManagement/README.md:1]: Don't like the readme inside the controller namespace. Resouces/doc folder should be usued for this

File: Controller/ContentManagement/CrudController.php [hunk 10]
```diff
@@ -358,4 +374,50 @@ public function getContentTypeInfo(ContentType $contentType)
                 'contentType' => $contentType,
         ]);
     }
+
+
+    /**
+     * @return array{username:string, displayName:string, roles:array<string>, email:string, circles:array<string>}
+     */
+    private function userToApiUserArray(User $user): array
```
Comment (Davidmattei) [inline Controller/ContentManagement/CrudController.php:382]: Small detail, public functions first private after. This should be moved to the bottom

File: Controller/ContentManagement/CrudController.php (hunk 6 above)
Comment (dameert) [reply Controller/ContentManagement/CrudController.php]: I do not get why $this->userService->getEnabledUsers() would expose too much information?
You are allready exposing this info in the controller (and the controller should not contain this logic). The service is part of your backend, why are you worried?

File: Controller/ContentManagement/CrudController.php (hunk 4 above)
Comment (dameert) [reply Controller/ContentManagement/CrudController.php]: I understand, but then make this function part of the userService. The controller should not have this logic.

File: Controller/ContentManagement/CrudController.php [hunk 11]
```diff
@@ -358,4 +374,50 @@ public function getContentTypeInfo(ContentType $contentType)
...
+        if (! $user->isEnabled()) {
+            throw new \RuntimeException('User disabled');
+        }
+
+        return $this->json($this->userToApiUserArray($user));
+    }
+
+    /**
+     * @Route("/api/user-profiles", defaults={"_format": "json"}, methods={"GET"})
+     * @IsGranted("ROLE_ADMIN")
```
Comment (coppee) [inline Controller/ContentManagement/CrudController.php:411]:
```suggestion
     * @IsGranted("ROLE_USER_MGMT", "ROLE_ADMIN")
```

File: Entity/User.php [hunk 12]
```diff
@@ -199,4 +199,16 @@ public function isEnabled();
      * @return string
      */
     public function getUsername();
+
```
Comment (dameert) [inline Entity/User.php:202]: Entity/User is now the User object from LocalUserBundle . The interface is renamed into Entity/UserInterface

File: Controller/ContentManagement/README.md (hunk 9 above)
Comment (dameert) [reply Controller/ContentManagement/README.md:1]: I agree with @Davidmattei this file should be part of the Resources/doc filder and be named "api.md" and have a reference from the main readme file. There is allready an api.md file over there.

File: Controller/ContentManagement/CrudController.php (hunk 11 above)
Comment (dameert) [reply Controller/ContentManagement/CrudController.php:411]: When did this role appear @coppee ?

File: Controller/ContentManagement/CrudController.php (hunk 10 above)
Comment (dameert) [reply Controller/ContentManagement/CrudController.php:382]: I still think this function should be part of the User object "$user->toApiArray()" The CrudController should not need to care about how to construct this array

File: Controller/ContentManagement/CrudController.php [hunk 13]
```diff
@@ -358,4 +374,50 @@ public function getContentTypeInfo(ContentType $contentType)
...
+    {
+        $user = $this->getUser();
+        if (! $user instanceof User) {
+            throw new \RuntimeException('User profile class not recognized');
+        }
+        if (! $user->isEnabled()) {
+            throw new \RuntimeException('User disabled');
+        }
+
+        return $this->json($this->userToApiUserArray($user));
```
Comment (dameert) [inline Controller/ContentManagement/CrudController.php:406]:
```suggestion
        return $this->json($user->toApiArray());
```

File: Controller/ContentManagement/CrudController.php (hunk 4 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: Moved in the service 9cbf75b7315e4fec4fa3edfeea82918a61565542

File: Controller/ContentManagement/CrudController.php (hunk 5 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: Done 9cbf75b7315e4fec4fa3edfeea82918a61565542

File: Controller/ContentManagement/CrudController.php (hunk 6 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: See 9cbf75b7315e4fec4fa3edfeea82918a61565542

File: Controller/ContentManagement/README.md (hunk 9 above)
Comment (theus77) [reply Controller/ContentManagement/README.md]: Moved to the resources folder dd143f01f6bc2da86899bfaaea3c65185ec43a35

File: Controller/ContentManagement/CrudController.php (hunk 10 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: moved c5d0e467c0d7786047dff7c40f41a1f4c26cca4c

File: Controller/ContentManagement/CrudController.php (hunk 11 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: It an incoming role

File: Entity/User.php (hunk 12 above)
Comment (theus77) [reply Entity/User.php:445]: The branch has been rebased

File: Controller/ContentManagement/CrudController.php (hunk 4 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: even better in the user entity fda1219a36c1bec8f610426fdb8efc358057a658

File: Controller/ContentManagement/CrudController.php (hunk 5 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: even better in the user entity fda1219a36c1bec8f610426fdb8efc358057a658

File: Controller/ContentManagement/CrudController.php (hunk 6 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: even better in the user entity fda1219a36c1bec8f610426fdb8efc358057a658

File: Controller/ContentManagement/CrudController.php (hunk 10 above)
Comment (theus77) [reply Controller/ContentManagement/CrudController.php]: even better in the user entity fda1219a36c1bec8f610426fdb8efc358057a658

File: Entity/User.php (hunk 12 above)
Comment (dameert) [reply Entity/User.php:461]: We see all the commits of develop in your PR, I think you needed to "merge" to not rewrite all the commit hashes!

**Comment by dameert** [comment]
Use squash!

**Pull request merged**